### DISCUSSION TITLE
Dimension of array requires constant and non parametric variability

### DISCUSSION THREAD

**Discussion by mpradovelasco** [OpenModelica/OpenModelica#11022] [Q&A] [answered]
Hi, 

In the following code:
```mo
package OuterTest
  parameter Integer n = 1;
  model usey
    outer Real x[n];
    Real y;
  equation
    y = x[1];
  end usey;
  //
  model Top
    inner Real x[n] = fill(2.4,n);
    usey A;
  end Top;
end OuterTest;
```
The model Top cannot be translated since OM gives the error:

[Component 'n' was found in an enclosing scope but is not a constant.](omeditmessagesbrowser:///CybSim.test?lineNumber=442)

I think this is a bug. The error dissapears if n is declared inside the models, or with constant variability.

System: OM latest nightly build 1.22-dev on Windows 11

Thank you for your help!

**Comment by Arinomo** [comment]
Maybe it has something to do with putting parameter on `package` level? 

normally you put a `parameter` inside a `model` class, create an instance of this model and you can change the parameter of said instance.

I don't know how parameter on `package` level was handled when an instance of model inside said `package` is created. But putting `parameter` inside a `package` does not make much sense, since you can not change it and  makes it a `constant` anyway.

I modified your example a little bit and use the `inner/outer` on `parameter n`. it seems to translate ok.

```.mo
model OuterTest
  inner parameter Integer n = 3;
  model usey
    outer parameter Integer n;
    outer Real x[n];
    Real y;
  equation
    y = x[n];
  end usey;
  //
  model Top
    outer parameter Integer n;
    inner Real x[n];
    usey A;
    equation
      for i in 1:n loop
        x[i] = 2.4*i;
      end for;
  end Top;
Top top;  
end OuterTest;
```

Edit to answer the title of your post:
The size of array need to be known during compile time. As the example above show, you CAN define a parameter and use it as array size but you can not change the size of the array after it. i.e if you re-simulate you model w/o recompile the model, then the array size will always constant (3 in above example) regardless what value you put in `n`

**Comment by mahge** [comment, marked as answer]
I think there are multiple issues here. 

First, a `package` is not allowed to contain anything other than class declarations and constant components. 
[Modelica Spec-3.6 Section 4.6](https://specification.modelica.org/maint/3.6/class-predefined-types-and-declarations.html#specialized-classes)
>> package – May only contain declarations of classes and constants. Enhanced to allow import of elements of packages. (See also [chapter 13](https://specification.modelica.org/maint/3.6/packages.html) on packages.)

Second, you are not allowed to use a component from an enclosing scope if it is not a constant. 
[Modelica Spec-3.6 Section 5.3.1](https://specification.modelica.org/maint/3.6/scoping-name-lookup-and-flattening.html#simple-name-lookup)

>> Reference to variables successfully looked up in an enclosing class is only allowed for variables declared as constant.


----------------------------------------------
@perost, @casella, @adrpo can probably tell you more.

**Reply by Arinomo** [reply]
>> Second, you are not allowed to use a component from an enclosing scope if it is not a constant.

Ah... i thought one can use inner and outer parameter in this case, but the example on [Modelica Spec-3.6 Section 5.5](https://specification.modelica.org/maint/3.6/scoping-name-lookup-and-flattening.html#simultaneous-inner-outer-declarations) does not use enclosing class. Using inner/outer in MWE from @okmechak makes more sense.

**Comment by okmechak** [comment]
Maybe something like this?

```modelica
package OuterTest
  model usey
    parameter Integer n = 1;
    outer Real x[n];
    Real y;
  equation
    y = x[1];
  end usey;
  
  model Top
    parameter Integer n = 1; 
    inner Real x[n] = fill(2.4,n);
    usey A(n = n); 
  end Top;
end OuterTest;
```

**Comment by mpradovelasco** [comment]
Hi,
Certainly, there are good solutions that avoid the use of a parameter in the enclosing scope, as you described. I was in doubt about the error because I did not know what specification of Modelica may explain the behavior. 

Comments from @mahge are particularly clear.
Thanks to all from your great help!
